dev/Microsoft.Process.Environment/EnvironmentManager.cpp

```diff
@@ -120,15 +115,15 @@ namespace winrt::Microsoft::ProjectReunion::implementation
             {
                 THROW_IF_FAILED(HRESULT_FROM_WIN32(RegDeleteValue(environmentVariableKey.get()
                     , name.c_str())));
```

Comment:
environmentVariableKey points to the registry location in either HKCU or HKLM for the environment variables.

I thought...it was a generally good idea to check in case a registry operation failed.

Reply:
Generally one should have rigorous trapping of all error paths.

But deletion - removal - sometimes isn't blocking. Here if you fail to delete the regkey it's fatal. Is that the right answer?

At a minimum this fails if the regkey doesn't exist (ERROR_FILE_NOT_FOUND or ERROR_PATH_NOT_FOUND). That's wrong. See other comments on the subject

Reply:
Not sure I agree with the suggested change (to swallow the failure). The developer is calling this API to effect a change, a deletion in this case. While I agree, the end-goal is the same, the operation is not. A developer may want to handle the case where a registry value is unexpectedly missing--perhaps for auditing or some sort of misconfiguration UX. The suggested change would make handling that case impossible and knock it out of alignment with Win32 Registry APIs.

Reply:
>Is that the right answer?

I didn't mean to suggest swallowing the error here. Just asking to think it through - is this required or best-effort?

But regardless the code needs changing, as ERROR_FILE_NOT_FOUND and ERROR_PATH_NOT_FOUND are error codes but not error conditions. You want the value gone, whether you delete it (ERROR_SUCCESS) or it's already gone (ERROR_FILE/PATH_NOT_FOUND) you get your wish.

Reply:
I don't understand. I have [beefbrain](https://hypnospace.fandom.com/wiki/Beefbrain). I think you are suggesting removal of THROW_IF_FAILED and effectively swallowing any non-success values. Am I missing something? Can you type out the exact change you're thinking about if that's not right?

Reply:
No. At a minimum I'm stating the current code is bugged

```
THROW_IF_FAILED(HRESULT_FROM_WIN32(RegDeleteValue(environmentVariableKey.get(), name.c_str())));
```

and needs to be

```
const auto rc{ RegDeleteValue(environmentVariableKey.get(), name.c_str()) };
THROW_HR_IF(HRESULT_FROM_WIN32(rc), (rc != ERROR_SUCCESS) && (rc != ERROR_FILE_NOT_FOUND));
```

Right now the code throws if the reg value doesn't exist. But this should be treated as a success case, not a failure.

This is an common annoyance of the registry APIs, equally commonly solved by making small wrappers -- OpenIfExists, DeleteIfExists, etc to map the common not-found cases to ERROR_SUCCESS/S_OK so the WIL macros neatly handle it

Reply:
Got it, thanks. This is an _internal_ delete occurring as a result of a developer calling `EnvironmentManager::SetEnvironmentVariable` with a blank value. Agree on delete-if-exists semantics here.

(I wonder if it makes more sense for this API to throw/error on blank input, and explicitly add a `EnvironmentManager::RemoveEnvironmentVariable` method. Perhaps out of scope.)

Reply:
RemoveEnvironmentVariable is a neat idea.  I want to keep the API as close to a Win32 API as possible to make the transition seamless.  Both C++ and C# have a SetEnvironmentVariable feature and the way to remove a variable is by setting the input to empty.